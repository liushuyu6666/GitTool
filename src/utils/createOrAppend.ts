export default function(container: Record<any, any []>, key: any, val: any): void {
  if (!Object.keys(container).includes(key)) {
    container[key] = [];
  }
  container[key].push(val);
}