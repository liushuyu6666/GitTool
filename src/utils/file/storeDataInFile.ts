import fs from 'fs';
import path from 'path';

export default function (filePath: string, data: any) {
  const json = JSON.stringify(data, null, 2);
  const lastSep = filePath.lastIndexOf(path.sep);
  const prefix = filePath.substring(0, lastSep);

  if (fs.existsSync(prefix)) {
    fs.writeFileSync(filePath, json);
    console.log(
      `[storeDataInFile]: data stored into ${path.join(
        filePath,
      )} successfully.`,
    );
  } else {
    throw new Error(
      `[file path error]: ${filePath} doesn\'t exist, cannot store data into that`,
    );
  }
}
