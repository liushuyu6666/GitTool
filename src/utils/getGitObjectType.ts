export enum GitObjectType {
  BLOB = 'blob',
  TREE = 'tree',
  COMMIT = 'commit',
  TAG = 'tag',
  BLOB_DELTA = 'blob_delta',
  TREE_DELTA = 'tree_delta',
  COMMIT_DELTA = 'commit_delta',
  TAG_DELTA = 'tag_delta',
  OFS_DELTA = 'ofs_delta',
  REF_DELTA = 'ref_delta',
  UNDEFINED = 'undefined',
}

export const GitPackObjectType: Record<number, GitObjectType> = {
  1: GitObjectType.COMMIT_DELTA,
  2: GitObjectType.TREE_DELTA,
  3: GitObjectType.BLOB_DELTA,
  4: GitObjectType.TAG_DELTA,
  5: GitObjectType.OFS_DELTA,
  6: GitObjectType.REF_DELTA,
};

export default function (type: string | number): GitObjectType {
  if (typeof type === 'string') {
    switch (type) {
      case GitObjectType.BLOB:
        return GitObjectType.BLOB;
      case GitObjectType.TREE:
        return GitObjectType.TREE;
      case GitObjectType.COMMIT:
        return GitObjectType.COMMIT;
      case GitObjectType.TAG:
        return GitObjectType.TAG;
      case GitObjectType.OFS_DELTA:
        return GitObjectType.OFS_DELTA;
      case GitObjectType.REF_DELTA:
        return GitObjectType.REF_DELTA;
      default:
        throw new Error(`Can\'t find the GitObjectType: ${type}`);
    }
  } else if (typeof type === 'number') {
    // if it is number, then it should come from pack file
    return GitPackObjectType[type];
  } else {
    throw new Error(`getGitObjectType error!`);
  }
}

const originalType: Array<GitObjectType> = [
  GitObjectType.BLOB,
  GitObjectType.TREE,
  GitObjectType.COMMIT,
  GitObjectType.TAG,
];

const originalDeltaType: Array<GitObjectType> = [
  GitObjectType.BLOB_DELTA,
  GitObjectType.TREE_DELTA,
  GitObjectType.COMMIT_DELTA,
  GitObjectType.TAG_DELTA,
]

const deltaType: Array<GitObjectType> = [
  GitObjectType.OFS_DELTA,
  GitObjectType.REF_DELTA,
];

export function isOriginalObject(type: GitObjectType): boolean {
  return originalType.includes(type);
}

export function isOriginalDeltaObject(type: GitObjectType): boolean {
  return originalDeltaType.includes(type);
}

export function isDeltaObject(type: GitObjectType): boolean {
  return deltaType.includes(type);
}
