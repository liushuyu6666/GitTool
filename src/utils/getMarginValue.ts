// TODO: should have a constant file to store this kind of value.
import { BYTE_POWER } from "../manipulateBuffer/ManipulateBuffer";

export default function (byteNumber: number) {
  if (byteNumber <= 1) return 0; 
  
  let sum = 0;
  for (let i = 0; i < byteNumber - 1; i++) {
    const octet = Math.pow(2, 7 - i);
    const newSeg = octet * Math.pow(BYTE_POWER, i);
    sum += newSeg;
  }

  return sum;
}

// 00100000 01000000 10000000