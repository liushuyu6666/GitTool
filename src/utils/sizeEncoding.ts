import { ManipulateBuffer } from "../manipulateBuffer/ManipulateBuffer";

export function parseSizeEncoding(chunk: Buffer, startIndex: number): [ManipulateBuffer, number] {
  const targetBuffer = chunk.subarray(startIndex);

  const manipulateBuffer = new ManipulateBuffer(false);
  let currIndex = -1;
  do {
    currIndex++;
    manipulateBuffer.fill(targetBuffer[currIndex], 1, 8);
  } while (targetBuffer[currIndex] > 0b10000000);
  manipulateBuffer.finish();

  return [manipulateBuffer, currIndex + 1];
}