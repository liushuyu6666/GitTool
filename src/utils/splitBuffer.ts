export default function (buffer: Buffer, delimiter=0x00) {
  const bufferArray: Buffer[] = [];

  let left: Buffer = buffer;
  while (left.indexOf(delimiter) !== -1) {
    const index = left.indexOf(delimiter);
    const former = left.subarray(0, index);
    left = left.subarray(index + 1);
    bufferArray.push(former);
  }
  bufferArray.push(left);

  return bufferArray;
}