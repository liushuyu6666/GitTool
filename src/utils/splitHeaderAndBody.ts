export default function (buffer: Buffer, delimiter=0x00): [Buffer, Buffer] {
  const idx = buffer.indexOf(delimiter);
  return [buffer.subarray(0, idx), buffer.subarray(idx + 1)];
}