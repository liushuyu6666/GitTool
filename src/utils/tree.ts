export interface GitTreeObjectEntry {
  mode: string;
  fileName: string;
  hex: string;
}
/**
 * 
 * @param entry The entry of the tree, format should be: 
 * <mode>0x20<fileName>0x00<20bytesHex>
 */
export function parseTreeEntryInPack(entry: Buffer): GitTreeObjectEntry {
  const first = entry.indexOf(0x20); // separate mode and others
  const second = entry.indexOf(0x00); // separate file name and hex

  const mode = entry.subarray(0, first).toString('ascii');
  const fileName = entry.subarray(first + 1, second).toString('ascii');
  const hex = entry.subarray(second + 1).toString('hex');

  return {
    mode,
    fileName,
    hex
  }
}