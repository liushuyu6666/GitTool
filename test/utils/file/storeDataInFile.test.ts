import fs from 'fs';
import storeDataInFile from '../../../src/utils/file/storeDataInFile';
import { GitObjectType } from '../../../src/utils/getGitObjectType';

describe('Test the storeDataInFile function:', () => {
  const data = {
    hash: 'test111111111111',
    type: GitObjectType.BLOB,
    size: 1,
    filePath: 'test_object',
    bodyOffsetStartIndex: 34,
    bodyOffsetEndIndex: 100,
  };

  test('Saving the data in the json file.', () => {
    const file = 'test/utils/file/data.json';

    // delete the file if exists
    if (fs.existsSync(file)) {
      fs.unlinkSync(file);
    }

    storeDataInFile(file, data);
    const csv = fs.readFileSync(file, 'utf-8');
    expect(csv).toMatchSnapshot();
  });

  test("If the folder doesn't exists, an error should be thrown.", () => {
    const file = 'test/not/exists';

    // delete the file if exists
    if (fs.existsSync(file)) {
      fs.unlinkSync(file);
    }

    expect(() => {
      storeDataInFile(file, data);
    }).toThrow(
      `[file path error]: ${file} doesn\'t exist, cannot store data into that`,
    );
  });
});
