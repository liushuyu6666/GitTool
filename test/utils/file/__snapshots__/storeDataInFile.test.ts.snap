// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the storeDataInFile function Saving the data in the json file 1`] = `
"{
  "hash": "test111111111111",
  "type": "blob",
  "size": 1,
  "filePath": "test_object",
  "bodyOffsetStartIndex": 34,
  "bodyOffsetEndIndex": 100
}"
`;

exports[`Test the storeDataInFile function: Saving the data in the json file. 1`] = `
"{
  "hash": "test111111111111",
  "type": "blob",
  "size": 1,
  "filePath": "test_object",
  "bodyOffsetStartIndex": 34,
  "bodyOffsetEndIndex": 100
}"
`;
